devices/src/virtio/vsock/csm/connection.rs

```diff
@@ -0,0 +1,632 @@
+            expiry: None,
+        }
+    }
+
+    /// Create a new host-initiated connection object.
+    ///
+    pub fn new_local_init(
+        stream: S,
+        local_cid: u64,
+        peer_cid: u64,
+        local_port: u32,
+        peer_port: u32,
+    ) -> Self {
+        Self {
+            local_cid,
+            peer_cid,
+            local_port,
+            peer_port,
+            stream,
+            state: ConnState::LocalInit,
+            tx_buf: TxBuf::new(),
+            fwd_cnt: Wrapping(0),
+            peer_buf_alloc: 0,
+            peer_fwd_cnt: Wrapping(0),
+            rx_cnt: Wrapping(0),
+            last_fwd_cnt_to_peer: Wrapping(0),
+            pending_rx: PendingRxSet::from(PendingRx::Request),
+            expiry: None,
+        }
+    }
+
+    /// Check if there is an expiry (kill) timer set for this connection, sometime in the future.
+    ///
+    pub fn will_expire(&self) -> bool {
+        match self.expiry {
+            None => false,
+            Some(t) => t > Instant::now(),
+        }
+    }
+
+    /// Check if this connection needs to be scheduled for forceful termination, due to its kill
+    /// timer having expired.
+    ///
+    pub fn has_expired(&self) -> bool {
+        match self.expiry {
+            None => false,
+            Some(t) => t <= Instant::now(),
+        }
+    }
+
+    /// Get the kill timer value, if one is set.
+    ///
+    pub fn expiry(&self) -> Option<Instant> {
+        self.expiry
+    }
+
+    /// Schedule the connection to be forcefully terminated ASAP (i.e. the next time the connection is
+    /// asked to yield a packet, via `recv_pkt()`).
+    ///
+    pub fn kill(&mut self) {
+        self.state = ConnState::Killed;
+        self.pending_rx.insert(PendingRx::Rst);
+    }
+
+    /// Send some raw data (a byte-slice) to the host host stream.
+    ///
+    /// Raw data can either be sent straight to the host stream, or to our TX buffer, if the former
+    /// fails.
+    ///
+    fn send_bytes(&mut self, buf: &[u8]) -> Result<()> {
+        if !self.tx_buf.is_empty() {
+            // If there is data in our TX buffer, we'll need to try and flush it out first.
+            // Flushing data out of the TX buffer also increases the "forwarded" bytes counter, since
+            // it's freed up some space for us.
+            self.fwd_cnt += Wrapping(self.tx_buf.flush_to(&mut self.stream)? as u32);
+            if self.tx_buf.is_empty() {
+                // Alright, we've drained the TX buffer; now, let's try this again, from the top.
+                self.send_bytes(buf)?;
+            } else {
+                // Couldn't drain the TX buffer, so the new data needs to go on top of the buffer.
+                self.tx_buf.push(buf)?;
+            }
+            return Ok(());
+        }
+
+        // The TX buffer is empty, so we can try to write straight to the host stream.
+        let written = match self.stream.write(buf) {
```

Comment:
I guess here goes the "write that returns 0" and `WouldBlock` chat again. In particular, if we keep acting like nothing really happened when we get 0 from write, or `WouldBlock` happens, is there any risk of generating a tight loop of such operations just because there's data to send, or we have logic which backs off and tries later somewhere?